### PR TITLE
Add oldest tag per release branch

### PR DESCRIPTION
**Pull request by rafiss**
This lets you easily see the oldest releases that have
a particular commit in them.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/8)

### CODE REVIEW

**Comment by jordanlewis** [comment]
This is amazing!!! I will leave it up to @rail to review but this is going to be a game changer (for engineering and other folks as well who are interested to see in which release a patch is available).

File: state.go [hunk 1]
```diff
@@ -321,12 +329,22 @@ func loadCommits(re repo, constraints ...string) (cs commits, err error) {
 			return commits{}, err
 		}
 		authorEmail := fields[3]
+		findTagArgs := []string{
+			"git", "-C", re.path(), "describe", "--match", "v[0-9]*", "--contains", sha.String(),
```
Comment (rafiss) [inline state.go:333]: one potential concern here: this is getting run for each commit on the branch. is there a way to get this info in O(1) instead?

**Comment by rail** [comment]
Let me see how it looks locally. It may take some time.

**Review by rail — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/8)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @jordanlewis and @rafiss)

___
*[state.go, line 333 at r1](https://reviewable.io/reviews/cockroachdb/backboard/8#-MbYjLBq3lJAZYit9ma1-r1-333:-MbbaszeBQekT1yqoOGR:btcdbua) ([raw file](https://github.com/cockroachdb/backboard/blob/981244e50e71ce06e04b40a9b870c375f5b5398a/state.go#L333)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

one potential concern here: this is getting run for each commit on the branch. is there a way to get this info in O(1) instead?
</blockquote></details>

So far it's been very slow for me. Can't even bootstrap it yet...

BTW, for some reason this change broke Ctrl+C, it won't exit without `kill -9`. Maybe because `capture()` doesn't use a cancellable context?

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/8)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @jordanlewis and @rail)

___
*[state.go, line 333 at r1](https://reviewable.io/reviews/cockroachdb/backboard/8#-MbYjLBq3lJAZYit9ma1-r1-333:-MbbzuBiDHFHCI4TxBuT:b-qmvu38) ([raw file](https://github.com/cockroachdb/backboard/blob/981244e50e71ce06e04b40a9b870c375f5b5398a/state.go#L333)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

So far it's been very slow for me. Can't even bootstrap it yet...

BTW, for some reason this change broke Ctrl+C, it won't exit without `kill -9`. Maybe because `capture()` doesn't use a cancellable context?
</blockquote></details>

ooo looks like git 2.32 just came out and has something that'll help

https://git-scm.com/docs/git-log#Documentation/git-log.txt-emdescribeoptionsem

```


    human-readable name, like git-describe[1]; empty string for undescribable commits. The describe string may be followed by a colon and zero or more comma-separated options. Descriptions can be inconsistent when tags are added or removed at the same time.

        match=<pattern>: Only consider tags matching the given glob(7) pattern, excluding the "refs/tags/" prefix.

        exclude=<pattern>: Do not consider tags matching the given glob(7) pattern, excluding the "refs/tags/" prefix.

```

which version of git is running on the backboard server?

**Review by rail — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/8)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @jordanlewis and @rafiss)

___
*[state.go, line 333 at r1](https://reviewable.io/reviews/cockroachdb/backboard/8#-MbYjLBq3lJAZYit9ma1-r1-333:-Mbc-gWsBgrm1v7bF-96:bjvleu4) ([raw file](https://github.com/cockroachdb/backboard/blob/981244e50e71ce06e04b40a9b870c375f5b5398a/state.go#L333)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ooo looks like git 2.32 just came out and has something that'll help

https://git-scm.com/docs/git-log#Documentation/git-log.txt-emdescribeoptionsem

```


    human-readable name, like git-describe[1]; empty string for undescribable commits. The describe string may be followed by a colon and zero or more comma-separated options. Descriptions can be inconsistent when tags are added or removed at the same time.

        match=<pattern>: Only consider tags matching the given glob(7) pattern, excluding the "refs/tags/" prefix.

        exclude=<pattern>: Do not consider tags matching the given glob(7) pattern, excluding the "refs/tags/" prefix.

```

which version of git is running on the backboard server?
</blockquote></details>

Let me check the current [Dockerfile](https://github.com/cockroachdb/backboard/blob/9005009ff4e12ca0a734932e610e286f19fdea3e/docker/Dockerfile#L1):
```
$docker run -ti golang:1.16 git --version
git version 2.20.1
```

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/8)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @jordanlewis and @rail)

___
*[state.go, line 333 at r1](https://reviewable.io/reviews/cockroachdb/backboard/8#-MbYjLBq3lJAZYit9ma1-r1-333:-MbcItKi8OveJkZNB5hb:b-g99tbc) ([raw file](https://github.com/cockroachdb/backboard/blob/981244e50e71ce06e04b40a9b870c375f5b5398a/state.go#L333)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Let me check the current [Dockerfile](https://github.com/cockroachdb/backboard/blob/9005009ff4e12ca0a734932e610e286f19fdea3e/docker/Dockerfile#L1):
```
$docker run -ti golang:1.16 git --version
git version 2.20.1
```
</blockquote></details>

ah gotcha. well actually, turns out that format syntax is equally slow.

but i figured out a faster way -- i'll just use the `%D` format verb, which shows the tags for the current commit. since we're scanning through all the commits linearly, we can just keep track of the "last seen tag" while iterating

**Review by rafiss — COMMENTED** [review]
Ready for another review

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/8#-:-MbdupIg1AhpyrH5dDCI:b-a0pjw2)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @jordanlewis and @rail)

**Review by rail — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/8)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @jordanlewis, @rafiss, and @rail)

___
*[state.go, line 308 at r2](https://reviewable.io/reviews/cockroachdb/backboard/8#-Mbg6aRM0JOor4UzLNE1:-Mbg6aRM0JOor4UzLNE2:b-vtfr5c) ([raw file](https://github.com/cockroachdb/backboard/blob/62f530159d0ffed4a17be625fa54c549f831e3fd/state.go#L308)):*
> ```Go
> func (a *acommit) OldestTags() string {
> 	if len(a.oldestTags) == 0 {
> 		return "none"
> ```

Maybe replace "none" with "" unless you meant to carry some meaning. Otherwise the UI looks a bit cluttered:


BTW, any idea why some "tags" are empty and some "none"?
___
*[state.go, line 341 at r2](https://reviewable.io/reviews/cockroachdb/backboard/8#-Mbg7-8X10ufu9YHd1cy:-Mbg7-8X10ufu9YHd1cz:b-37b3bk) ([raw file](https://github.com/cockroachdb/backboard/blob/62f530159d0ffed4a17be625fa54c549f831e3fd/state.go#L341)):*
> ```Go
> 		}
> 		authorEmail := fields[3]
> 		refNames := strings.Split(fields[4], ", ")
> ```

I believe this should be `fileds[5]`.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/8)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @jordanlewis and @rail)

___
*[state.go, line 308 at r2](https://reviewable.io/reviews/cockroachdb/backboard/8#-Mbg6aRM0JOor4UzLNE1:-MbgdK3c2V-AFXfjXR2i:b-7bbvga) ([raw file](https://github.com/cockroachdb/backboard/blob/62f530159d0ffed4a17be625fa54c549f831e3fd/state.go#L308)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Maybe replace "none" with "" unless you meant to carry some meaning. Otherwise the UI looks a bit cluttered:


BTW, any idea why some "tags" are empty and some "none"?

</blockquote></details>

hm, no empty vs "none" is confusing. i'll work on debugging that locally.

my idea with using "none" was to try to make it clear that this commit does not exist in any released versions yet. but it seems like it just ended up being more confusing. do you think i should just leave it as "" or would something like "not released yet" still be too cluttered?
___
*[state.go, line 341 at r2](https://reviewable.io/reviews/cockroachdb/backboard/8#-Mbg7-8X10ufu9YHd1cy:-MbgdPuQC58mHcy_4nhb:b-d11h3z) ([raw file](https://github.com/cockroachdb/backboard/blob/62f530159d0ffed4a17be625fa54c549f831e3fd/state.go#L341)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

I believe this should be `fileds[5]`.
</blockquote></details>

whoops! thanks

**Review by rail — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/8)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @jordanlewis, @rafiss, and @rail)

___
*[state.go, line 308 at r2](https://reviewable.io/reviews/cockroachdb/backboard/8#-Mbg6aRM0JOor4UzLNE1:-Mbge91f9Z4C58QDmuFf:brzkv47) ([raw file](https://github.com/cockroachdb/backboard/blob/62f530159d0ffed4a17be625fa54c549f831e3fd/state.go#L308)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm, no empty vs "none" is confusing. i'll work on debugging that locally.

my idea with using "none" was to try to make it clear that this commit does not exist in any released versions yet. but it seems like it just ended up being more confusing. do you think i should just leave it as "" or would something like "not released yet" still be too cluttered?
</blockquote></details>

I'd say, let's leave it empty for now and figure out what we can we do to improve the UX?

**Review by rafiss — COMMENTED** [review]
ready for one more review! i ran it locally and it seems better.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/8#-:-Mbh4EO5EIKgfCEOK6XM:bq05jpd)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @jordanlewis and @rail)

___
*[state.go, line 308 at r2](https://reviewable.io/reviews/cockroachdb/backboard/8#-Mbg6aRM0JOor4UzLNE1:-Mbh44xAAiPG0_bGdml1:brzn9gz) ([raw file](https://github.com/cockroachdb/backboard/blob/62f530159d0ffed4a17be625fa54c549f831e3fd/state.go#L308)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

I'd say, let's leave it empty for now and figure out what we can we do to improve the UX?
</blockquote></details>

sounds good; it's empty now
___
*[state.go, line 341 at r2](https://reviewable.io/reviews/cockroachdb/backboard/8#-Mbg7-8X10ufu9YHd1cy:-Mbh4B8JCOsQcfvJDWzT:bzc16vi) ([raw file](https://github.com/cockroachdb/backboard/blob/62f530159d0ffed4a17be625fa54c549f831e3fd/state.go#L341)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

whoops! thanks
</blockquote></details>

this resolved the weird "" vs "none" behavior too

**Review by rail — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/backboard/8#-:-MbilZEQ9GGH1ZkLWAYy:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @jordanlewis and @rafiss)

**Comment by rail** [comment]
It's live now! \o/

**Comment by jordanlewis** [comment]
This is amazing!

One question - how come there's no 20.2 tags in the tags column? I'd expect to see at least a couple... 🤔

**Comment by rafiss** [comment]
@jordanlewis I think it's because to see the 20.2 tags, you need to change the "base branch" to release-20.2. it only checks for backport commits for the branch that is currently selected.

this behavior is not entirely intentional -- if this seems confusing, i could add more logic so that it gets the tags from all release branches.

**Comment by jordanlewis** [comment]
I'm envisioning that we can use this dashboard to solve the problem of CS and TSE wanting to know whether a commit has been released yet and if so to which tag. Definitely scope creep, but I feel like we're so close. Doing what you say would get us there!

**Comment by rafiss** [comment]
I agree! here is the change https://github.com/cockroachdb/backboard/pull/9

**Pull request merged**